vault/plugin_catalog.go

```diff
@@ -59,14 +102,208 @@ func (c *Core) setupPluginCatalog(ctx context.Context) error {
 	return nil
 }
 
+type pluginClientConn struct {
+	*grpc.ClientConn
+	id string
+}
+
+var _ grpc.ClientConnInterface = &pluginClientConn{}
+
+func (d *pluginClientConn) Invoke(ctx context.Context, method string, args interface{}, reply interface{}, opts ...grpc.CallOption) error {
+	// Inject ID to the context
+	md := metadata.Pairs(pluginutil.MultiplexingCtxKey, d.id)
+	idCtx := metadata.NewOutgoingContext(ctx, md)
+
+	return d.ClientConn.Invoke(idCtx, method, args, reply, opts...)
+}
+
+func (d *pluginClientConn) NewStream(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	// Inject ID to the context
+	md := metadata.Pairs(pluginutil.MultiplexingCtxKey, d.id)
+	idCtx := metadata.NewOutgoingContext(ctx, md)
+
+	return d.ClientConn.NewStream(idCtx, desc, method, opts...)
+}
+
+func (p *pluginClient) Conn() grpc.ClientConnInterface {
+	return p.clientConn
+}
+
+// MultiplexingSupport determines if a plugin client supports multiplexing
+func (p *pluginClient) MultiplexingSupport() (bool, error) {
+	if p.client == nil {
+		return false, fmt.Errorf("plugin client is nil")
+	}
+	return pluginutil.MultiplexingSupport(p.client.NegotiatedVersion()), nil
+}
+
+// Close calls the plugin client's cleanupFunc to do any necessary cleanup on
+// the plugin client and the PluginCatalog. This implements the
+// plugin.ClientProtocol interface.
+func (p *pluginClient) Close() error {
+	p.logger.Debug("cleaning up plugin client connection", "id", p.id)
+	return p.cleanupFunc()
+}
+
+func (c *PluginCatalog) removePluginClient(name, id string) error {
+	var err error
+	extPlugin, ok := c.externalPlugins[name]
+	if !ok {
+		return fmt.Errorf("plugin client not found")
+	}
+
+	pluginClient := extPlugin.connections[id]
+	multiplexingSupport, err := pluginClient.MultiplexingSupport()
```

Comment:
Instead of re-computing this here, could we store in the extPlugin object if we are multiplexing or not? 

That would solve this issue: https://github.com/hashicorp/vault/pull/14060/files#r806345285